### PR TITLE
[ROCm] bugfixing conv_parameters and activated respective test

### PR DESCRIPTION
**Pull request by pemeliya**
This is bugfix in conv_parameters.cc originally appeared in https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/2000 and then in the follow-up PR: https://github.com/tensorflow/tensorflow/pull/61941

As requested, I am submitting this bugfix as a separate PR. 
I also had to extend the tests for Eigen::half datatype and 'Tanh' and 'Sigmoid' activations
since in this case fused matmul uses the CuDNN fallback which, in its turn, 
uses MatmulParameters constructor (from tensorflow/core/util/autotune_maps/conv_parameters.cc).

Additionally, I also removed MatmulParameters class from tensorflow/core/kernels/matmul_op.h which is a deadcode (to remove any confusions).

@akuegel: could you please review this as this is a follow up to https://github.com/tensorflow/tensorflow/pull/61941 PR ?

### CODE REVIEW

File: tensorflow/core/kernels/matmul_op_test.cc
Comment (akuegel) [inline tensorflow/core/kernels/matmul_op_test.cc:261]: nit: seems this line is left-over from debugging?

File: tensorflow/core/kernels/matmul_op_test.cc
Comment (akuegel) [inline tensorflow/core/kernels/matmul_op_test.cc:227]: Would it be possible to also pass a rtol value instead of increasing the atol value to 1e-1 (which seems rather big)?

File: tensorflow/core/kernels/matmul_op_test.cc
Comment (akuegel) [inline tensorflow/core/kernels/matmul_op_test.cc:358]: nit: {} around the if body.

File: tensorflow/core/kernels/matmul_op_test.cc
Comment (akuegel) [inline tensorflow/core/kernels/matmul_op_test.cc:369]: nit: {} around the if body.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/matmul_op_test.cc
Comment (pemeliya) [reply tensorflow/core/kernels/matmul_op_test.cc:261]: yes, this can be removed

File: tensorflow/core/kernels/matmul_op_test.cc
Comment (pemeliya) [reply tensorflow/core/kernels/matmul_op_test.cc:227]: yes, the error seems to be rather big for Eigen::half datatype, especially for the test Matmul256x128x64.
Then I need to investigate this further, so far I can only reduce the 'atol' to 1e-2.
```
i = 0 Tx[i] = 2.0625 Ty[i] = 2.060546875
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (4.296875 not close to 4.30078125)
Expected: true
i = 85 Tx[i] = 4.296875 Ty[i] = 4.30078125
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (4.359375 not close to 4.36328125)
Expected: true
i = 96 Tx[i] = 4.359375 Ty[i] = 4.36328125
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (4.421875 not close to 4.42578125)
Expected: true
i = 181 Tx[i] = 4.421875 Ty[i] = 4.42578125
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (2.12109375 not close to 2.119140625)
Expected: true
i = 187 Tx[i] = 2.12109375 Ty[i] = 2.119140625
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (2.078125 not close to 2.076171875)
Expected: true
i = 229 Tx[i] = 2.078125 Ty[i] = 2.076171875
tensorflow/core/framework/tensor_testutil.cc:184: Failure
Value of: IsClose(Tx[i], Ty[i], typed_atol, typed_rtol)
  Actual: false (4.5078125 not close to 4.50390625)
Expected: true
i = 243 Tx[i] = 4.5078125 Ty[i] = 4.50390625
```
I have also played around this 'rtol' values but could not figure out a meanigful value which works fine for all cases

Comment (akuegel) [reply tensorflow/core/kernels/matmul_op_test.cc:227]: So I guess a combination of higher atol + specifying a similar rtol value should likely give the best result.
Can you try for example atol=1e-3 rtol=1e-3?

Comment (pemeliya) [reply tensorflow/core/kernels/matmul_op_test.cc:227]: yes, surpringly, with these values of atol/rtol as you suggested, the test for Eigen::half also works fine (though it takes ages because CUDNN does some autotuning first to find the proper algorithm to use):
```
[----------] 8 tests from Test/FusedMatMulWithBiasOpTest/1, where TypeParam = Eigen::half
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x128x64
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x128x64 (149 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x256
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x256 (131 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x256x1
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x256x1 (89 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x1
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x1 (19 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x128x64WithActivation
2023-10-05 10:14:04.231568: I external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:440] Loaded cuDNN version 8904
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x128x64WithActivation (220203 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x256WithActivation
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x256WithActivation (46056 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x256x1WithActivation
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul256x256x1WithActivation (0 ms)
[ RUN      ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x1WithActivation
[       OK ] Test/FusedMatMulWithBiasOpTest/1.MatMul1x256x1WithActivation (0 ms)
[----------] 8 tests from Test/FusedMatMulWithBiasOpTest/1 (266651 ms total)

[----------] Global test environment tear-down
[==========] 16 tests from 2 test suites ran. (269672 ms total)
[  PASSED  ] 16 tests.
```

**Review by akuegel — APPROVED** [review]

**Comment by akuegel** [comment]
It seems there are 2 failing tests:

```
[  FAILED  ] [Test/FusedMatMulWithBiasOpTest/1].MatMul256x128x64WithActivation, where TypeParam = Eigen::half
[  FAILED  ] [Test/FusedMatMulWithBiasOpTest/1].MatMul1x256x256WithActivation, where TypeParam = Eigen::half
```

**Comment by pemeliya** [comment]
> It seems there are 2 failing tests:
> 
> ```
> [  FAILED  ] [Test/FusedMatMulWithBiasOpTest/1].MatMul256x128x64WithActivation, where TypeParam = Eigen::half
> [  FAILED  ] [Test/FusedMatMulWithBiasOpTest/1].MatMul1x256x256WithActivation, where TypeParam = Eigen::half
> ```
> 
> And the errors look like:
> 
> `NOT_FOUND: No algorithm worked!`

OK, I will have a look at it tomorrow, possibly it depends on CUDNN version and GPU architecture (I have tested on SM 7.0 GPU). It says that CUDNN cannot not find a suitable mamul algorithm for the problem.

**Comment by pemeliya** [comment]
I would wait until the related PR is merged to upstream: https://github.com/tensorflow/tensorflow/pull/61941
because it applies some modifications to fused_matmul_op, then I can rebase and try running it once again.

**Review by akuegel — APPROVED** [review]

**Review by akuegel — COMMENTED** [review]
FYI: I am using a different fix internally, we can just avoid create Eigen::half test cases on CPU.

**Comment by pemeliya** [comment]
> FYI: I am using a different fix internally, we can just avoid create Eigen::half test cases on CPU.

Yes, if you have a better approach how to skip these tests, can you share it with me ?
Basically I wanted to skip those tests for which no OpKernel is available but it seems that GTEST_SKIP() does not break the test execution immediately, so there are some "patches" to make it working..

**Comment by pemeliya** [comment]
@akuegel: do you know why this PR has been rejected ? I last updated it over the weekend and was hoping that the tests will run fine now..

**Comment by akuegel** [comment]
Can you explain what you mean with rejected? This got merged 46 minutes ago. It took a while because I needed internal approvals from US-based people, so they happened after my working hours. And this morning I was OOO as well.

**Comment by pemeliya** [comment]
ah sorry, I was confused by this message: 
```
[PR Queue] automation moved this from Approved by Reviewer to Closed/Rejected [1 hour ago]
```
many thanks for the help !

**Comment by akuegel** [comment]
It seems my idea of a fix doesn't work, because we seem to run the _cpu version of the test also with --config=cuda. Some followup might be needed, maybe also a temporary revert.

**Comment by pemeliya** [comment]
OK, then I guess the only way is to check for the errors at runtime.. 

For example, in Python there are exceptions, and some tests do it in this way:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/nn_ops/depthwise_conv_op_base.py#L377-L387

In C++, I also tried to selectively skip subtests depending on tsl::Status() message but without exceptions it's more cumbersome..

**Comment by akuegel** [comment]
I think I found a solution, adding the tag no_cuda_on_cpu_tap should avoid that the test is automatically run on CPU with --config=cuda. That should be good enough.

**Pull request closed without merging**
